pkg/skaffold/watch/changes.go

```diff
@@ -18,59 +18,86 @@ package watch
 
 import (
 	"os"
+	"sort"
 	"time"
 
 	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
 )
 
-type fileMap struct {
-	count        int
-	lastModified time.Time
-}
+type fileMap map[string]time.Time
 
+// TODO(mrick): cached tree extension ala git
 func stat(deps func() ([]string, error)) (fileMap, error) {
+	state := fileMap{}
 	paths, err := deps()
 	if err != nil {
-		return fileMap{}, errors.Wrap(err, "listing files")
+		return state, errors.Wrap(err, "listing files")
 	}
-
-	last, err := lastModified(paths)
-	if err != nil {
-		return fileMap{}, err
+	for _, path := range paths {
+		stat, err := os.Stat(path)
+		if err != nil {
+			return nil, errors.Wrapf(err, "unable to stat file %s", path)
+		}
+		state[path] = stat.ModTime()
 	}
 
-	return fileMap{
-		count:        len(paths),
-		lastModified: last,
-	}, nil
+	return state, nil
 }
 
-func lastModified(paths []string) (time.Time, error) {
-	var last time.Time
+type Events struct {
+	Added    []string
+	Modified []string
+	Deleted  []string
+}
 
-	for _, path := range paths {
-		stat, err := os.Stat(path)
-		if err != nil {
-			if os.IsNotExist(err) {
-				continue // Ignore files that don't exist
-			}
+func (e Events) HasChanged() bool {
+	added, deleted, modified := len(e.Added), len(e.Deleted), len(e.Modified)
+	if added > 0 {
+		logrus.Debugf("[watch event] added: %s", e.Added)
```

Comment:
having the logs in a getter is a bit surprising (nothing prevents a client of this API would call HasChanged multiple times, but it can generate unnecessary clutter in the logs) I'd rather have them in the `events()` at calculation time.

Reply:
Good call!  I’ll change